### PR TITLE
Post-commit updates for bpo-33073.

### PR DESCRIPTION
**Pull request by serhiy-storchaka**
* Simplify the C code.
* Simplify tests and make them more strict and robust.
* Add references in the documentation.

https://bugs.python.org/issue33073

### CODE REVIEW

File: Lib/test/test_long.py
Comment (pablogsal) [inline Lib/test/test_long.py:1355]: Question: what about the `enum.IntEnum` tests? (I assume those can be derived from an int subclass but I just want to confirm).

Comment (serhiy-storchaka) [reply Lib/test/test_long.py:1355]: `enum.IntEnum` is just an `int` subclass.

**Review by pablogsal — APPROVED** [review]

**Comment by serhiy-storchaka** [comment]
@lisroach, please take a look.

**Comment by vstinner** [comment]
The change LGTM, but got no activity after @pablogsal approved it one month ago. I decided to merge it, since I consider that it enhances the code and I don't want this PR to be forgotten.

I changed the commit title, but kept its description.

Thanks @serhiy-storchaka, I like new tests :-)

cc @lisroach

**Pull request merged**
